krun/mail.py

```diff
@@ -18,33 +18,47 @@
 
 
 class Mailer(object):
-    def __init__(self, recipients, max_mails):
-        self.recipients = recipients
+    def __init__(self, recipients=None, max_mails=5):
+        if recipients is not None:
+            self.recipients = recipients
+        else:
+            self.recipients = []
         self.hostname = socket.gethostname()
         self.short_hostname = self.hostname.split(".")[0]
 
-        # After sending max_mails emails, we stop sending more so as
-        # not to spam. Some emails however, you will always want to send.
+        # After sending the maximum number of emails, we stop sending more so
+        # as not to spam. Some emails however, you will always want to send.
         # For these use send(..., bypass_limiter=True).
         self.max_mails = max_mails
-        self.n_mails_sent = 0
 
     def set_recipients(self, recipients):
         self.recipients = recipients
 
     def _wrap_para(self, txt):
         return "\n".join(textwrap.wrap(txt, WRAP_THRESHOLD))
 
-    def send(self, append_subject, inner_body, bypass_limiter=False):
+    def send(self, append_subject, inner_body, bypass_limiter=False,
+             manifest=None):
+        if manifest is not None:
+            num_mails_sent = manifest.num_mails_sent
+        else:
+            # It's OK to call this without a manifest (e.g. outside the
+            # scheduler loop, where there is no manifest to speak of), but
+            # without a manifest we can't know how many emails have been sent.
+            # So, the only time this is OK is if we are skipping the limiter
+            # anyway.
+            assert bypass_limiter  # Krun can't know how many mails were sent
```

Comment:
Depending where it is called, it may go via a couple of intermediate handlers, but yes, it will arrive at the top level, as for all AssertionErrors. We wouldn't want to comment every single assertion, so what makes this one special?